### PR TITLE
Edit Profile and Stimulus Modal

### PR DESCRIPTION
**Pull request by Shaka-n**
~~WIP Reason: along the way I broke the form! but it looks nice. Fixing shortly~~

Adds an 'edit' profile form! 
Introduces ViewComponent for reusable component views! 
Adds the modal component!
Fixes the logout bug!

Update: Still needs tests, but ready to go otherwise.

### CODE REVIEW

File: app/components/modal_component.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<%= turbo_frame_tag "modal" do %>
+  <%= tag.div id:'modal-overlay',
+              class: 'fixed z-20 w-screen h-screen bg-gray-100/75',
+              data: { controller: 'modal',
+                      modal_target: 'modal',
+                      action: 'turbo:submit-end->modal#submitEnd keydown.esc@window->modal#closeWithEscape mousedown@window->modal#closeWithBackgroundClick'} do %>
```
Comment (emmahyde) [inline app/components/modal_component.html.erb:6]: 🤮

File: app/models/profile.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@
 #
 class Profile < ApplicationRecord
   belongs_to :user
+  accepts_nested_attributes_for :user
```
Comment (emmahyde) [inline app/models/profile.rb:15]: 👌

File: app/controllers/profiles_controller.rb [hunk 3]
```diff
@@ -34,15 +34,10 @@ def edit
...
-      @user.update(user_params)
-      @profile.update(profile_params)
-    end
-
     respond_to do |format|
       format.html do
-        if update_result
-          redirect_to profile_path(@user)
+        if @profile.update!(profile_params)
+          redirect_to profile_path(@user.unique_name), notice: 'Profile was successfully updated.'
```
Comment (emmahyde) [inline app/controllers/profiles_controller.rb:40]: During the rewiring I did to make things navigate to the unique name, I defined `to_param` on the `User` model:
https://github.com/emmahyde/grouper/blob/8d4c3cb5cff26ae82de3355448d971d448a3cc98/app/models/user.rb#L46-L48

And also on the `Profile` cuz they're basically the same thing:
https://github.com/emmahyde/grouper/blob/f15e2ed3f0b0e1dcfb5d5b2a17ec48e43dd7240b/app/models/profile.rb#L21-L23

This is what all these Rails autogenerated `_path` helpers call under the hood when an ActiveRecord object is passed. The default behavior takes the id off of the object, so I've customized that behavior to pluck the `unique_name` instead.

Long story short, I think you can just do
```suggestion
          redirect_to profile_path(@user), notice: 'Profile was successfully updated.'
```

it will call `to_param` if you don't specify!

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -13,6 +13,7 @@
   </head>
 
   <body>
+    <%= turbo_frame_tag 'modal'%>
```
Comment (emmahyde) [inline app/views/layouts/application.html.erb:16]:
```suggestion
    <%= turbo_frame_tag 'modal' %>
```

File: app/views/profiles/edit.html.erb [hunk 5]
```diff
@@ -1,34 +1,3 @@
...
-      <% end %>
-      
-      <div>
-        <%= f.submit "Save Changes", class:"block text-white bg-gray-800 rounded-full px-2" %>
-      </div>
-    <% end %>
-    </div>
-  </div>
+<%= render ModalComponent.new(title: "Editing Profile") do %>
+  <%= render "form", profile: @profile, user: @user %>
```
Comment (emmahyde) [inline app/views/profiles/edit.html.erb:2]: omfg

File: app/views/profiles/show.html.erb [hunk 6]
```diff
@@ -14,7 +12,7 @@
         <div id='friend-request-button' class='absolute -top-4 -right-2 '>
           <span class='bg-blue-200 py-1 px-2 rounded-full opacity-50'>Request Friend</span>
           <% if current_user == @user %>
-            <%= link_to 'Edit', edit_profile_path(@user), id: 'edit-profile-button', class: 'bg-blue-200 py-2 px-2 rounded-full w-9', data: { turbo_frame: 'edit-profile-modal' } %>
```
Comment (emmahyde) [inline app/views/profiles/show.html.erb:17]: the previous thing I mentioned is why this navigates via the unique_name, same thing

**Review by emmahyde — APPROVED** [review]

**Comment by emmahyde** [comment]
Honestly inspiring, I was def trying to do too many things at once. This is an awesome incremental change

File: app/components/modal_component.html.erb (hunk 1 above)
Comment (Shaka-n) [reply app/components/modal_component.html.erb:6]: 🤣 I know, it looks awful, and the connection with the actual controller actions is so tenuous. 

I was thinking of writing a helper module for stuff like this to make it more readable. It would be really cool if we could make something like Phoenix's [verified_routes](https://www.germanvelasco.com/blog/phoenix-1-7-verified-routes). That way we wouldn't need to have magic strings. We could round up all the controller actions in the JS at compile time and set a global verified_stimulus_actions map. The tough part would be somehow generating all the valid event triggers to go with them... maybe we wouldn't have to though. Ending up with something like this:
```ruby
$verified_stimulus_actions = { 'modal': ['closeEnd', 'submitEnd'] }
```
... then have a macro/resolver of some kind check if the action exists and fail more visibly if it does not:
```ruby
module StimulusHelper
  def verify_action (action)
    check_for_action($verified_stimulus_actions)
  end
end
```
...Something like that anyway. 

In the meantime, I was thinking something like this just for readability:

```ruby
module StimulusHelper
  def action_list(actions)
    actions.join(' ')
  end
end

...
### modal_component.html.erb
<tag.div id:'modal-overlay;
  data: {
    controller: 'modal',
    action: StimulusHelper.action_list(["turbo:submit-end->modal#submitEnd", "keydown.esc@window->modal#closeWithEscape"])
  } do %>
  <%= content %>
  <% end %>
```

File: app/controllers/profiles_controller.rb (hunk 3 above)
Comment (Shaka-n) [reply app/controllers/profiles_controller.rb:40]: ahhhhhhhh, that makes so much sense! I was having trouble understanding that change. Ty!

File: app/components/modal_component.html.erb (hunk 1 above)
Comment (Shaka-n) [reply app/components/modal_component.html.erb:6]: Gonna make an issue for this actually. Don't want to hold up this PR any longer but the more I think about it the better an idea this seems. Maybe even a gem worth publishing, if I can't find one that does something similar 🤔
